File: Android Car Dealer Tracker/app/src/main/java/com/ezcar24/business/util/calculator/VehicleFinancialsCalculator.kt
package com.ezcar24.business.util.calculator

import com.ezcar24.business.data.local.Expense
import com.ezcar24.business.data.local.ExpenseCategoryType
import com.ezcar24.business.data.local.Vehicle
import java.math.BigDecimal
import java.math.RoundingMode

object VehicleFinancialsCalculator {

    private const val INTERMEDIATE_SCALE = 4
    private const val DISPLAY_SCALE = 2
    private const val PERCENTAGE_SCALE = 2

    fun calculateTotalCost(
        vehicle: Vehicle,
        allExpenses: List<Expense>,
        holdingCost: BigDecimal
    ): BigDecimal {
        val purchasePrice = vehicle.purchasePrice

        val expensesTotal = allExpenses
            .filter { it.deletedAt == null }
            .map { it.amount }
            .fold(BigDecimal.ZERO) { acc, amount -> acc.add(amount) }

        return purchasePrice
            .add(expensesTotal)
            .add(holdingCost)
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateTotalExpensesOnly(
        vehicle: Vehicle,
        allExpenses: List<Expense>
    ): BigDecimal {
        val expensesTotal = allExpenses
            .filter { it.deletedAt == null }
            .map { it.amount }
            .fold(BigDecimal.ZERO) { acc, amount -> acc.add(amount) }

        return vehicle.purchasePrice
            .add(expensesTotal)
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateROI(salePrice: BigDecimal, totalCost: BigDecimal): BigDecimal? {
        if (totalCost.compareTo(BigDecimal.ZERO) == 0) {
            return null
        }

        val profit = salePrice.subtract(totalCost)
        return profit
            .divide(totalCost, INTERMEDIATE_SCALE, RoundingMode.HALF_UP)
            .multiply(BigDecimal(100))
            .setScale(PERCENTAGE_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateProfitEstimate(askingPrice: BigDecimal, totalCost: BigDecimal): BigDecimal? {
        return askingPrice
            .subtract(totalCost)
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateActualProfit(salePrice: BigDecimal, totalCost: BigDecimal): BigDecimal {
        return salePrice
            .subtract(totalCost)
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateHoldingCostPercentage(
        holdingCost: BigDecimal,
        totalCost: BigDecimal
    ): BigDecimal? {
        if (totalCost.compareTo(BigDecimal.ZERO) == 0) {
            return null
        }

        return holdingCost
            .divide(totalCost, INTERMEDIATE_SCALE, RoundingMode.HALF_UP)
            .multiply(BigDecimal(100))
            .setScale(PERCENTAGE_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateExpenseBreakdown(expenses: List<Expense>): Map<ExpenseCategoryType, BigDecimal> {
        return expenses
            .filter { it.deletedAt == null }
            .groupBy { it.expenseType }
            .mapValues { (_, expenses) ->
                expenses
                    .map { it.amount }
                    .fold(BigDecimal.ZERO) { acc, amount -> acc.add(amount) }
                    .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
            }
    }

    fun calculateBreakEvenPrice(
        vehicle: Vehicle,
        allExpenses: List<Expense>,
        holdingCost: BigDecimal,
        targetROI: BigDecimal = BigDecimal.ZERO
    ): BigDecimal {
        val totalCost = calculateTotalCost(vehicle, allExpenses, holdingCost)

        if (targetROI.compareTo(BigDecimal.ZERO) == 0) {
            return totalCost
        }

        val roiMultiplier = targetROI
            .divide(BigDecimal(100), INTERMEDIATE_SCALE, RoundingMode.HALF_UP)
            .add(BigDecimal.ONE)

        return totalCost
            .multiply(roiMultiplier)
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateRecommendedAskingPrice(
        vehicle: Vehicle,
        allExpenses: List<Expense>,
        holdingCost: BigDecimal,
        targetROI: BigDecimal = BigDecimal("20.00")
    ): BigDecimal {
        return calculateBreakEvenPrice(vehicle, allExpenses, holdingCost, targetROI)
    }

    fun isProfitable(salePrice: BigDecimal, totalCost: BigDecimal): Boolean {
        return salePrice.compareTo(totalCost) > 0
    }

    fun getProfitStatus(
        salePrice: BigDecimal,
        totalCost: BigDecimal
    ): ProfitStatus {
        val profit = salePrice.subtract(totalCost)
        val roi = calculateROI(salePrice, totalCost)

        return when {
            profit.compareTo(BigDecimal.ZERO) < 0 -> ProfitStatus.LOSS
            profit.compareTo(BigDecimal.ZERO) == 0 -> ProfitStatus.BREAK_EVEN
            roi != null && roi.compareTo(BigDecimal("20")) >= 0 -> ProfitStatus.HIGH_PROFIT
            else -> ProfitStatus.PROFIT
        }
    }

    enum class ProfitStatus {
        LOSS,
        BREAK_EVEN,
        PROFIT,
        HIGH_PROFIT
    }
}

File: iOS Car Dealer Tracker/Ezcar24Business/Services/Calculators/VehicleFinancialsCalculator.swift
import Foundation

class VehicleFinancialsCalculator {
    
    private static let intermediateScale = 4
    private static let displayScale = 2
    private static let percentageScale = 2
    
    static func calculateTotalCost(
        vehicle: Vehicle,
        expenses: [Expense],
        holdingCost: Decimal
    ) -> Decimal {
        let purchasePrice = vehicle.purchasePrice?.decimalValue ?? 0
        
        let expensesTotal = expenses
            .filter { $0.deletedAt == nil }
            .reduce(Decimal(0)) { $0 + ($1.amount?.decimalValue ?? 0) }
        
        return (purchasePrice + expensesTotal + holdingCost).rounded(to: displayScale)
    }
    
    static func calculateTotalExpensesOnly(
        vehicle: Vehicle,
        expenses: [Expense]
    ) -> Decimal {
        let expensesTotal = expenses
            .filter { $0.deletedAt == nil }
            .reduce(Decimal(0)) { $0 + ($1.amount?.decimalValue ?? 0) }
        
        return ((vehicle.purchasePrice?.decimalValue ?? 0) + expensesTotal).rounded(to: displayScale)
    }
    
    static func calculateROI(
        salePrice: Decimal,
        totalCost: Decimal
    ) -> Decimal? {
        guard totalCost != 0 else {
            return nil
        }
        
        let profit = salePrice - totalCost
        return ((profit / totalCost) * 100).rounded(to: percentageScale)
    }
    
    static func calculateProfitEstimate(
        askingPrice: Decimal,
        totalCost: Decimal
    ) -> Decimal? {
        return (askingPrice - totalCost).rounded(to: displayScale)
    }
    
    static func calculateActualProfit(
        salePrice: Decimal,
        totalCost: Decimal
    ) -> Decimal {
        return (salePrice - totalCost).rounded(to: displayScale)
    }
    
    static func calculateBreakEvenPrice(
        totalCost: Decimal,
        targetROI: Decimal = 0
    ) -> Decimal {
        guard targetROI != 0 else {
            return totalCost
        }
        
        let roiMultiplier = (targetROI / 100) + 1
        return (totalCost * roiMultiplier).rounded(to: displayScale)
    }
    
    static func calculateHoldingCostPercentage(
        holdingCost: Decimal,
        totalCost: Decimal
    ) -> Decimal? {
        guard totalCost != 0 else {
            return nil
        }
        
        return ((holdingCost / totalCost) * 100).rounded(to: percentageScale)
    }
    
    static func calculateExpenseBreakdown(
        expenses: [Expense]
    ) -> [ExpenseCategoryType: Decimal] {
        return expenses
            .filter { $0.deletedAt == nil }
            .compactMap { expense -> (ExpenseCategoryType, Decimal)? in
                guard let categoryType = expense.categoryTypeEnum else { return nil }
                return (categoryType, expense.amount?.decimalValue ?? 0)
            }
            .reduce(into: [:]) { result, pair in
                result[pair.0, default: 0] += pair.1
            }
            .mapValues { $0.rounded(to: displayScale) }
    }
    
    static func calculateRecommendedAskingPrice(
        vehicle: Vehicle,
        expenses: [Expense],
        holdingCost: Decimal,
        targetROI: Decimal = 20
    ) -> Decimal {
        let totalCost = calculateTotalCost(vehicle: vehicle, expenses: expenses, holdingCost: holdingCost)
        return calculateBreakEvenPrice(totalCost: totalCost, targetROI: targetROI)
    }
    
    static func isProfitable(salePrice: Decimal, totalCost: Decimal) -> Bool {
        return salePrice > totalCost
    }
    
    static func getProfitStatus(
        salePrice: Decimal,
        totalCost: Decimal
    ) -> ProfitStatus {
        let profit = salePrice - totalCost
        let roi = calculateROI(salePrice: salePrice, totalCost: totalCost)
        
        if profit < 0 {
            return .loss
        } else if profit == 0 {
            return .breakEven
        } else if let roi = roi, roi >= 20 {
            return .highProfit
        } else {
            return .profit
        }
    }
}

enum ProfitStatus {
    case loss
    case breakEven
    case profit
    case highProfit
    
    var displayName: String {
        switch self {
        case .loss:
            return "Loss"
        case .breakEven:
            return "Break Even"
        case .profit:
            return "Profit"
        case .highProfit:
            return "High Profit"
        }
    }
    
    var colorName: String {
        switch self {
        case .loss:
            return "red"
        case .breakEven:
            return "yellow"
        case .profit:
            return "green"
        case .highProfit:
            return "blue"
        }
    }
}

private extension Decimal {
    func rounded(to scale: Int) -> Decimal {
        var result = Decimal()
        var localCopy = self
        NSDecimalRound(&result, &localCopy, scale, .plain)
        return result
    }
}
